File: app/src/main/java/torille/fi/lurkforreddit/data/RedditService.java
package torille.fi.lurkforreddit.data;

import io.reactivex.Observable;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import torille.fi.lurkforreddit.data.models.jsonResponses.MultiredditListing;
import torille.fi.lurkforreddit.data.models.jsonResponses.PostListing;
import torille.fi.lurkforreddit.data.models.jsonResponses.RedditToken;
import torille.fi.lurkforreddit.data.models.jsonResponses.SubredditChildren;
import torille.fi.lurkforreddit.data.models.jsonResponses.SubredditListing;

/**
 * Created by eva on 22.4.2017.
 */

public interface RedditService {

    interface Auth {
        @FormUrlEncoded
        @POST("access_token")
        Observable<RedditToken> getAuthToken(@Field("grant_type") String grant_type, @Field("device_id") String UUID);

        @FormUrlEncoded
        @POST("access_token")
        Observable<RedditToken> getUserAuthToken(@Field("grant_type") String grant_type, @Field("code") String code, @Field("redirect_uri") String redirectUri);

        @FormUrlEncoded
        @POST("access_token")
        Observable<RedditToken> refreshUserToken(@Field("grant_type") String grant_type, @Field("refresh_token") String token);

    }

    interface Reddit {
        @GET("subreddits/mine/subscriber")
        Observable<SubredditListing> getMySubreddits(@Query("limit") int count);

        @GET("subreddits/default")
        Observable<SubredditListing> getDefaultSubreddits(@Query("limit") int count);

        @GET("/api/multi/mine")
        Observable<MultiredditListing[]> getUserMultireddits();

        @GET("me/m/{multipath}")
        Call<ResponseBody> getMultiRedditData(@Path(value = "multipath", encoded = true) String multiPath);

        @GET("me/m/{multipath}")
        Call<ResponseBody> getMultiRedditDataNextPage(@Path(value = "multipath", encoded = true) String multiPath, @Query("after") String afterPage);

        @GET("{subreddit}")
        Observable<PostListing> getSubreddit(@Path(value = "subreddit", encoded = true) String subredditUrl);

        @GET("{subreddit}")
        Observable<PostListing> getSubredditNextPage(@Path(value = "subreddit", encoded = true) String subredditUrl, @Query("after") String afterPageId);

        @GET("{comments}")
        Observable<ResponseBody> getComments(@Path(value = "comments", encoded = true) String commentUrl);

        @GET("api/morechildren")
        Observable<ResponseBody> getMoreComments(@Query(value = "link_id") String parentId, @Query(value = "children") String childId, @Query(value = "api_type") String json);

        @GET("subreddits/search")
        Observable<SubredditListing> searchSubreddits(@Query(value = "q") String searchQuery, @Query(value = "sort") String sortBy);

        @GET("subreddits/search")
        Observable<SubredditListing> searchSubredditsNextPage(@Query(value = "q") String searchQuery, @Query(value = "sort") String sortBy, @Query(value = "after") String after);

        @GET("{subreddit}/about")
        Call<SubredditChildren> getSubredditInfo(@Path(value = "subreddit", encoded = true) String subredditName);

    }
}

File: app/src/main/java/torille/fi/lurkforreddit/data/RedditService.kt
package torille.fi.lurkforreddit.data

import io.reactivex.Flowable
import io.reactivex.Observable
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.http.*
import torille.fi.lurkforreddit.data.models.jsonResponses.*

/**
 * API endpoints for Reddit API
 */

interface RedditService {

    interface Auth {
        @FormUrlEncoded
        @POST("access_token")
        fun getAuthToken(@Field("grant_type") grant_type: String, @Field("device_id") UUID: String): Observable<RedditToken>

        @FormUrlEncoded
        @POST("access_token")
        fun getUserAuthToken(
            @Field("grant_type") grant_type: String, @Field("code") code: String, @Field(
                "redirect_uri"
            ) redirectUri: String
        ): Observable<RedditToken>

        @FormUrlEncoded
        @POST("access_token")
        fun refreshUserToken(@Field("grant_type") grant_type: String, @Field("refresh_token") token: String): Observable<RedditToken>

    }

    interface Reddit {
        @GET("subreddits/mine/subscriber")
        fun getMySubreddits(@Query("limit") count: Int): Flowable<SubredditListing>

        @GET("subreddits/default")
        fun getDefaultSubreddits(@Query("limit") count: Int): Flowable<SubredditListing>

        @GET("/api/multi/mine")
        fun getUserMultireddits(): Flowable<Array<MultiredditListing>>

        @GET("me/m/{multipath}")
        fun getMultiRedditData(
            @Path(
                value = "multipath",
                encoded = true
            ) multiPath: String
        ): Call<ResponseBody>

        @GET("me/m/{multipath}")
        fun getMultiRedditDataNextPage(
            @Path(
                value = "multipath",
                encoded = true
            ) multiPath: String, @Query("after") afterPage: String
        ): Call<ResponseBody>

        @GET("{subreddit}")
        fun getSubreddit(
            @Path(
                value = "subreddit",
                encoded = true
            ) subredditUrl: String
        ): Observable<PostListing>

        @GET("{subreddit}")
        fun getSubredditNextPage(
            @Path(
                value = "subreddit",
                encoded = true
            ) subredditUrl: String, @Query("after") afterPageId: String
        ): Observable<PostListing>

        @GET("{comments}")
        fun getComments(
            @Path(
                value = "comments",
                encoded = true
            ) commentUrl: String
        ): Observable<ResponseBody>

        @GET("api/morechildren")
        fun getMoreComments(
            @Query(value = "link_id") parentId: String, @Query(value = "children") childId: String, @Query(
                value = "api_type"
            ) json: String
        ): Observable<ResponseBody>

        @GET("subreddits/search")
        fun searchSubreddits(@Query(value = "q") searchQuery: String, @Query(value = "sort") sortBy: String): Flowable<SubredditListing>

        @GET("subreddits/search")
        fun searchSubredditsNextPage(
            @Query(value = "q") searchQuery: String, @Query(value = "sort") sortBy: String, @Query(
                value = "after"
            ) after: String
        ): Flowable<SubredditListing>

        @GET("{subreddit}/about")
        fun getSubredditInfo(
            @Path(
                value = "subreddit",
                encoded = true
            ) subredditName: String
        ): Flowable<SubredditChildren>

    }
}
